### ISSUE TITLE
CloseAplication.pas suggestion

### ISSUE THREAD

**Issue by ijbranch** [corneliusdavid/ccLib#2]
It occurred to me that CloseApplication.pas could have two new events..
1.  OnWarn - Triggers just prior to the warning dialog.  I would use this to turn off in App timers and/or flush buffers.
2.   OnResume - Triggers if the Users elects not to terminate.  I would use this to restart in App timers, etc.
Regards,
Ian

**Comment by corneliusdavid** [comment]
Those are good suggestions but I don't time or immediate use for them so they'll be shelved for a while. If you'd like to fork this project and create a pull-request, I'll incorporate the changes.

Thanks, David.

**Comment by ijbranch** [comment]
Hi David,
              Sorry, I don't know how to fork and all that stuff, just do a full download using SVN.
              I will make the changes as I see them and submit like the last for your assessment.

Regards,
Ian

**Comment by corneliusdavid** [comment]
OK. If you ever want to get into that, check this out: [first commit](https://first-commit.netlify.app). It's where I started.

**Comment by ijbranch** [comment]
Tks David.  I really should but just haven't had the need.
Is there any risk of corrupting the originals??

**Comment by corneliusdavid** [comment]
No. When you create a PR, it's a request to add a commit of your changes to my repository. I review them before making them a permanent part of the repository. 

Furthermore, commits are always additive--they don't overwrite previous commits (there are exceptions to this), they just overwrite the files in the repository. That means you can always go back to a previous version of the files by checking out a previous commit.

**Comment by ijbranch** [comment]
Cheers.

**Comment by ijbranch** [comment]
Hi David,
              Got a pull done and am now playing with ccLib in my GitHib area.
              I have added the extra events, well I thought I had..
{code}
type
  TOnApplicationMsg = procedure(var Msg: TMsg; var Handled: Boolean) of object;

  [ComponentPlatforms(pfidWindows)]
  TCloseApplication = class(TComponent)
    procedure IdleTimerTimer(Sender: TObject);
  private
    FCountDownScreenDisplayed: Boolean; // Var To Prevent Proc CloseAppBecauseIdleTimeExceeded From Being Called Recursively
    FMinutesAppAllowedToBeIdle: Integer; // Number of Minutes App Allowed To Be Idle
    FSecondsPromptedOnShutdown: Integer; // Number of Seconds Prevent Shutdown Screen Is Displayed Before App Shutdown Occurs
    FTimerInterval: Integer; // Interval Of IdleTimer (Should Be 60000 to simulate 1 minute)
    FPreventCancelEditInserts: Boolean; // Flag To Prevent Cancellation of Edits & Inserts On Shutdown
    IdleTimer: TTimer;
    FShowAppName: Boolean;

    procedure CloseAppBecauseIdleTimeExceeded;
    procedure CancelEditsInsertsInDatamodules;
    procedure CancelEditsInsertsInOpenForms;

    procedure SetMinutesAppAllowedToBeIdle(Minutes: Integer);
    procedure SetSecondsPromptedOnShutdown(Seconds: Integer);
  protected
    procedure Loaded; override;
  public
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
  published
    property MinutesAppAllowedToBeIdle: Integer read FMinutesAppAllowedToBeIdle write SetMinutesAppAllowedToBeIdle default 60;
    property PreventCancelEditInserts: Boolean read FPreventCancelEditInserts write FPreventCancelEditInserts default False;
    property SecondsPromptedOnShutdown: Integer read FSecondsPromptedOnShutdown write SetSecondsPromptedOnShutdown default 90;
    property ShowAppName: Boolean read FShowAppName write FShowAppName default False;
    property TimerInterval: Integer read FTimerInterval write FTimerInterval default 60000;

    property OnAppTermination: TNotifyEvent read FOnAppTermination write FOnAppTermination;
    property OnWarning: TNotifyEvent read FOnWarning write FOnWarning;
    property OnResume: TNotifyEvent read FOnResume write FOnResume;

  end;
{code}
              But only the OnAppTermination Event shows in the component.
              I have rebuilt the ccLib library.
              Is there somewhere else I need to address?

Regards & TIA,
Ian

**Comment by ijbranch** [comment]
Corrected code.  Cut & Paste error - Finger fault..
{code}
type
  TOnApplicationMsg = procedure(var Msg: TMsg; var Handled: Boolean) of object;

  [ComponentPlatforms(pfidWindows)]
  TCloseApplication = class(TComponent)
    procedure IdleTimerTimer(Sender: TObject);
  private
    FCountDownScreenDisplayed: Boolean; // Var To Prevent Proc CloseAppBecauseIdleTimeExceeded From Being Called Recursively
    FMinutesAppAllowedToBeIdle: Integer; // Number of Minutes App Allowed To Be Idle
    FSecondsPromptedOnShutdown: Integer; // Number of Seconds Prevent Shutdown Screen Is Displayed Before App Shutdown Occurs
    FTimerInterval: Integer; // Interval Of IdleTimer (Should Be 60000 to simulate 1 minute)

    FOnAppTermination: TNotifyEvent; // Pointer To App Termination Event (if Defined)
    FOnWarning: TNotifyEvent; // Pointer To On Warning Event (if Defined)
    FOnResume: TNotifyEvent; // Pointer To On Resume Event (if Defined)

    FPreventCancelEditInserts: Boolean; // Flag To Prevent Cancellation of Edits & Inserts On Shutdown
    IdleTimer: TTimer;
    FShowAppName: Boolean;

    procedure CloseAppBecauseIdleTimeExceeded;
    procedure CancelEditsInsertsInDatamodules;
    procedure CancelEditsInsertsInOpenForms;

    procedure SetMinutesAppAllowedToBeIdle(Minutes: Integer);
    procedure SetSecondsPromptedOnShutdown(Seconds: Integer);
  protected
    procedure Loaded; override;
  public
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
  published
    property MinutesAppAllowedToBeIdle: Integer read FMinutesAppAllowedToBeIdle write SetMinutesAppAllowedToBeIdle default 60;
    property PreventCancelEditInserts: Boolean read FPreventCancelEditInserts write FPreventCancelEditInserts default False;
    property SecondsPromptedOnShutdown: Integer read FSecondsPromptedOnShutdown write SetSecondsPromptedOnShutdown default 90;
    property ShowAppName: Boolean read FShowAppName write FShowAppName default False;
    property TimerInterval: Integer read FTimerInterval write FTimerInterval default 60000;

    property OnAppTermination: TNotifyEvent read FOnAppTermination write FOnAppTermination;
    property OnWarning: TNotifyEvent read FOnWarning write FOnWarning;
    property OnResume: TNotifyEvent read FOnResume write FOnResume;

  end;
{Code}

**Comment by ijbranch** [comment]
Issue resolved.  Mixing 32 & 64 bit doesn’t work. ☹   Finger fault.

**Comment by corneliusdavid** [comment]
Sorry it took so long to respond but I've just added the two requested events and a demo app that demonstrates everything in the `CloseApplication` component.

**Comment by ijbranch** [comment]
Tks David,
I can delete my branch now. :-)

Ian

**Comment by ijbranch** [comment]
Hi David,
I added a timer countdown into my AppIdleWarn... :-)
[CloseApplication.zip](https://github.com/corneliusdavid/ccLib/files/9237866/CloseApplication.zip)
Feel free to use.
